nativelib/src/main/resources/gc/immix/Allocator.c

```diff
@@ -0,0 +1,188 @@
+#include <stdlib.h>
+#include "Allocator.h"
+#include "Line.h"
+#include "Block.h"
+#include "stats/AllocatorStats.h"
+#include <stdio.h>
+#include <memory.h>
+
+BlockHeader *getNextBlock(Allocator *allocator);
+bool getNextLine(Allocator *allocator);
+
+Allocator *Allocator_create(word_t *heapStart, int blockCount) {
+    Allocator *allocator = malloc(sizeof(Allocator));
+    allocator->heapStart = heapStart;
+
+    BlockList_init(&allocator->recycledBlocks, heapStart);
+    BlockList_init(&allocator->freeBlocks, heapStart);
+
+    // Init the free block list
+    allocator->freeBlocks.first = (BlockHeader *)heapStart;
+    BlockHeader *lastBlockHeader =
+        (BlockHeader *)(heapStart + ((blockCount - 1) * WORDS_IN_BLOCK));
+    allocator->freeBlocks.last = lastBlockHeader;
+    lastBlockHeader->header.nextBlock = LAST_BLOCK;
+
+// Block stats
+#ifdef ALLOCATOR_STATS
+    allocator->stats = AllocatorStats_create();
+    allocator->stats->blockCount = (uint64_t)blockCount;
+#endif
+
+    Allocator_initCursors(allocator);
+
+    return allocator;
+}
+
+bool Allocator_initCursors(Allocator *allocator) {
+
+    // Init cursor
+    allocator->block = NULL;
+    allocator->cursor = NULL;
+    allocator->limit = NULL;
+
+    getNextLine(allocator);
+
+    // Init large cursor
+    if (BlockList_isEmpty(&allocator->freeBlocks)) {
+        return false;
+    }
+    BlockHeader *largeHeader =
+        BlockList_removeFirstBlock(&allocator->freeBlocks);
+    allocator->largeBlock = largeHeader;
+    allocator->largeCursor = Block_getFirstWord(largeHeader);
+    allocator->largeLimit = Block_getBlockEnd(largeHeader);
+
+    return true;
+}
+
+/*
+ * Overflow allocation uses only free blocks, it is used when the bump limit of
+ * the fast allocator is too small to fit
+ * the block to alloc.
+ */
+word_t *overflowAllocation(Allocator *allocator, size_t size) {
+    word_t *start = allocator->largeCursor;
+    word_t *end = (word_t *)((uint8_t *)start + size);
+
+    if (end > allocator->largeLimit) {
+        // DIFFERENT FROM IMMIX
```

Comment:
This comment should be removed, it's different from the mmtk code because of multithreading, the access to the global allocator needs to be synchronised.